planner/cascades/implementation_rules.go

```diff
@@ -446,22 +480,48 @@ func (r *ImplHashJoinBuildRight) Match(expr *memo.GroupExpr, prop *property.Phys
 }
 
 // OnImplement implements ImplementationRule OnImplement interface.
-func (r *ImplHashJoinBuildRight) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+func (r *ImplHashJoinBuildRight) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) ([]memo.Implementation, error) {
 	join := expr.ExprNode.(*plannercore.LogicalJoin)
+	var joinImpls []memo.Implementation
 	switch join.JoinType {
 	case plannercore.SemiJoin, plannercore.AntiSemiJoin,
 		plannercore.LeftOuterSemiJoin, plannercore.AntiLeftOuterSemiJoin:
-		return getImplForHashJoin(expr, reqProp, 1, false), nil
+		joinImpls = append(joinImpls, getImplForHashJoin(expr, reqProp, 1, false))
+		return joinImpls, nil
 	case plannercore.InnerJoin:
-		return getImplForHashJoin(expr, reqProp, 1, false), nil
+		joinImpls = append(joinImpls, getImplForHashJoin(expr, reqProp, 1, false))
+		return joinImpls, nil
 	case plannercore.LeftOuterJoin:
-		return getImplForHashJoin(expr, reqProp, 1, false), nil
+		joinImpls = append(joinImpls, getImplForHashJoin(expr, reqProp, 1, false))
+		return joinImpls, nil
 	case plannercore.RightOuterJoin:
-		return getImplForHashJoin(expr, reqProp, 0, true), nil
+		joinImpls = append(joinImpls, getImplForHashJoin(expr, reqProp, 0, true))
+		return joinImpls, nil
 	}
 	return nil, nil
 }
 
+// ImplMergeJoin implements LogicalMergeJoin to PhysicalMergeJoin.
+type ImplMergeJoin struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplMergeJoin) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	return true
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplMergeJoin) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) ([]memo.Implementation, error) {
+	join := expr.ExprNode.(*plannercore.LogicalJoin)
+	physicalMergeJoins := join.GetMergeJoin(reqProp)
```

Comment:
We cannot directly use `join.GetMergeJoin` here. It will call `join.Schema()` inside this method.